garden-backend-service/src/api/search/utils.py

```diff
@@ -0,0 +1,113 @@
+        ```
+
+        This will generate a query with `WHERE` clauses matching the title and tags.
+    """
+    for filter in filters:
+        if not hasattr(model, filter.field_name):
+            raise ValueError(f"Invalid filter field_name: {filter.field_name}")
+        for value in filter.values:
+            if type(getattr(model, filter.field_name).type) is ARRAY:
+                stmt = stmt.where(
+                    func.array_to_string(getattr(model, filter.field_name), " ").match(
+                        value
+                    )
+                )
+            else:
+                stmt = stmt.where(
+                    func.cast(getattr(model, filter.field_name), TEXT).match(value)
+                )
+    return stmt
+
+
+async def calculate_facets(db: AsyncSession, query: Select) -> GardenSearchFacets:
+    """Calculate and return search facets for a given query.
+
+    This function computes facet counts for tags, authors, and years based on the
+    result set from the provided SQLAlchemy Select statement. Facets are calculated
+    by grouping on specific fields (`tags`, `authors`, and `year`) and counting
+    occurrences within the subset of gardens returned by the query.
+
+    Args:
+        db (AsyncSession): The asynchronous SQLAlchemy session used to execute queries.
+        query (Select): A SQLAlchemy `Select` statement representing the filtered subset
+                        of gardens to calculate facets from.
+
+    Returns:
+        GardenSearchFacets: An instance of `GardenSearchFacets` containing three fields:
+            - `tags` (dict[str, int]): A dictionary where keys are individual tags and values
+              are the count of gardens associated with each tag.
+            - `authors` (dict[str, int]): A dictionary where keys are author names and values
+              are the count of gardens authored by each individual.
+            - `year` (dict[str, int]): A dictionary where keys are years (as strings) and values
+              are the count of gardens created in each year.
+
+    Note:
+        This function assumes that the query provided returns results from a table or view that
+        includes the columns `tags`, `authors`, and `year` as arrays or individual fields.
+    """
+    filtered_gardens = query.subquery("filtered_gardens")
+
+    tags_query = select(
+        func.unnest(filtered_gardens.c.tags).label("tag"), func.count().label("count")
+    ).group_by(column("tag"))
+
+    authors_query = select(
+        func.unnest(filtered_gardens.c.authors).label("author"),
+        func.count().label("count"),
+    ).group_by(column("author"))
+
+    year_query = select(
+        filtered_gardens.c.year.label("year"), func.count().label("count")
+    ).group_by(column("year"))
+
+    tags_result = await db.execute(tags_query)
+    authors_result = await db.execute(authors_query)
+    year_result = await db.execute(year_query)
```

Comment:
I had the same thought and tried this running these in an `asyncio.gather`. It needed some extra setup to work because the AsyncSession object is [mutable and not thread safe](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-faq-threadsafe). Basically, we need a unique session for each parallel query which is not too hard. I was able to get it working, but issuing the queries in parallel doubled average response time in my benchmark vs issuing them serially.

Reply:
Interesting! Thanks for digging in to that